ecosystem/typescript/sdk/src/transaction_builder/builder_utils.ts

```diff
@@ -72,84 +72,108 @@ export function ensureBigInt(val: number | bigint | string): bigint {
+    if (argVal instanceof Uint8Array) {
+      serializer.serializeBytes(argVal);
+      return;
     }
-    addr.serialize(serializer);
-    return;
-  }
-  if (argType instanceof TypeTagVector) {
-    // We are serializing a vector<u8>
-    if (argType.value instanceof TypeTagU8) {
-      if (argVal instanceof Uint8Array) {
-        serializer.serializeBytes(argVal);
-        return;
-      }
-
-      if (typeof argVal === "string") {
-        serializer.serializeStr(argVal);
-        return;
-      }
+    if (argVal instanceof HexString) {
+      serializer.serializeBytes(argVal.toUint8Array());
+      return;
     }
-
-    if (!Array.isArray(argVal)) {
-      throw new Error("Invalid vector args.");
+    if (typeof argVal === "string") {
+      serializer.serializeStr(argVal);
+      return;
     }
+    // If it isn't any of those types, then it must just be an actual array of numbers
+  }
 
-    serializer.serializeU32AsUleb128(argVal.length);
-
-    argVal.forEach((arg) => serializeArg(arg, argType.value, serializer));
-    return;
+  if (!Array.isArray(argVal)) {
+    throw new Error("Invalid vector args.");
   }
 
-  if (argType instanceof TypeTagStruct) {
-    const { address, module_name: moduleName, name } = (argType as TypeTagStruct).value;
-    if (
-      `${HexString.fromUint8Array(address.address).toShortString()}::${moduleName.value}::${name.value}` !==
-      "0x1::string::String"
-    ) {
-      throw new Error("The only supported struct arg is of type 0x1::string::String");
-    }
-    assertType(argVal, ["string"]);
+  serializer.serializeU32AsUleb128(argVal.length);
 
+  argVal.forEach((arg) => serializeArgInner(arg, argType.value, serializer, depth + 1));
+}
+
+function serializeStruct(argVal: any, argType: TypeTag, serializer: Serializer, depth: number) {
+  const { address, module_name: moduleName, name, type_args: typeArgs } = (argType as TypeTagStruct).value;
+  const structType = `${HexString.fromUint8Array(address.address).toShortString()}::${moduleName.value}::${name.value}`;
+  if (structType === "0x1::string::String") {
+    assertType(argVal, ["string"]);
     serializer.serializeStr(argVal);
-    return;
+  } else if (structType === "0x1::object::Object") {
+    serializeAddress(argVal, serializer);
+  } else if (structType === "0x1::option::Option") {
+    if (typeArgs.length !== 1) {
+      throw new Error(`Option has the wrong number of type arguments ${typeArgs.length}`);
+    }
+    serializeOption(argVal, typeArgs[0], serializer, depth);
+  } else {
+    throw new Error("Unsupported struct type in function argument");
+  }
+}
+
+function serializeOption(argVal: any, argType: TypeTag, serializer: Serializer, depth: number) {
+  // For option, we determine if it's empty or not empty first
+  // empty option is nothing, we specifically check for undefined to prevent fuzzy matching
+  if (argVal === undefined) {
+    serializer.serializeU32AsUleb128(0);
+  } else {
+    // Something means we need an array of 1
```

Comment:
it might be good to define primitive types of bcs / move somewhere ... and have common test vectors... the bcs upstream repo is missing this... and considering our specialization I'm not convinced it is the best route to go.